Spectre mitigation on heap access overflow checks.

cfallin opened the pull request:
This PR adds a conditional move following a heap bounds check through which the address to be accessed flows. This conditional move ensures that even if the branch is mispredicted (access is actually out of bounds, but speculation goes down in-bounds path), the acually accessed address is zero (a NULL pointer) rather than the out-of-bounds address.

The technique used is well-known and fairly standard: see, e.g., [SpiderMonkey's heap access bounds check](https://searchfox.org/mozilla-central/source/js/src/jit/arm64/MacroAssembler-arm64-inl.h#1726-1729) (thanks to @bnjbvr for locating this).

The mitigation is controlled by a flag that is off by default, but can be set by the embedding, to avoid unexpected performance impact for existing users.

Note that the mitigation is unneccessary when we use the "huge heap" technique on 64-bit systems, in which we allocate a range of virtual address space such that no 32-bit offset can reach other data. Hence, this only affects small-heap configurations.

The generated code is not great on aarch64 because the overflow flag must be materialized as a boolean in a register, rather than reusing flags (because the dataflow crosses basic blocks); in theory we could do much better with a later lowering, codegenning one comparison and two ops that use its flags.

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

bnjbvr commented on cranelift/codegen/meta/src/shared/settings.rs:
```diff
@@ -264,5 +264,13 @@ pub(crate) fn define() -> SettingGroup {
+    settings.add_bool(
+        "enable_heap_access_spectre_mitigation",
+        "Enable Spectre mitigation on heap bounds checks.",
+        false,
```
The risk of causing harm by disabling Spectre mitigations by default seems much worse than the performance hit that will be caused by its being set all the time, so I think we should enable it by default, explicit the performance overhead here, and say that while it's enabled by default, there are certain environments in which it's fine to disable it.

cfallin replied:
See below: will need to be `false` by default unless we also implement for the legacy x86 backend (or until we delete it), which (unless I'm mistaken) seems to be missing some of the necessary bits for cmovs.

cfallin replied:
Now on by default with latest version.

bnjbvr commented on cranelift/codegen/src/legalizer/heap.rs:
```diff
@@ -198,5 +220,11 @@ fn compute_addr(
+    if let Some(c) = spectre_oob_cond {
+        let final_addr = pos.ins().iadd(base, offset);
+        let zero = pos.ins().iconst(addr_ty, 0);
+        pos.func.dfg.replace(inst).select(c, zero, final_addr);
```
Generating IR for the Spectre check seems quite dangerous, at this point; any optimization could reorganize the code (move it around in the best case, delete it in the worst case), so I don't think this is the right solution here.

bjorn3 replied:
Even worse, there is a legalization for `select` that uses branches when there is no encoding that uses `cmov`:

https://github.com/bytecodealliance/wasmtime/blob/c9a3f05afd45961b0b397f97c4ad79cd7a7c807d/cranelift/codegen/src/legalizer/mod.rs#L476-L513

cfallin replied:
Changed to use a specific, effectful instruction that the optimizer and legalizations do not know about.

julian-seward1 commented:
(Speaking from a position of partial ignorance) it would be best to do this at the aarch64 emission point.  I mean, add an `AArch64::Inst` case with whatever stuff we need, including the conditional move.  Then we're guaranteed that no CLIR or isel transformation can break it, and also that regalloc won't insert any reloads in between the branch and cmov.

I suspect that's not feasible, because it would require a combined conditional-branch-and-cmov `Inst`, and branches are handled specially in the newBE.  So the next best option is implement this at the isel level, I guess.

cfallin commented:
These are all good points -- I hadn't considered that an effect-free cmov might be hoisted above the branch, for example, or legalized back into a form that gets speculated on.

I like @bnjbvr's suggestion the best here: a `spectre_icmp` that is a black box to optimization passes would stay in place on the correct path, and we could guarantee its lowering. Perhaps we could call it something like `spectre_guard_cmov`...

cfallin commented:
Updated to use a `select_spectre_guard` instruction, which is like `select` except that it is (i) marked as effectful, and (ii) unknown to any of the optimization passes or legalizations. On AArch64, this lowers to the same conditional move.

Unfortunately, unless I'm mistaken, on x86 (legacy backend) it seems that full support for `cmov` instructions is missing? E.g., there is no legalization for select -> selectif, only `expand_select` that @bjorn3 pointed out (that always lowers to a CFG diamond); and in `.../x86/opcodes.rs`, there is only the opcode for `CMOV_OVERFLOW`. Perhaps I'm missing some interaction that results in full cmov support. However, for now, I have not added support for the Spectre-specific guard to the legacy x86 backend. Given our policy of putting any x86-specific efforts toward the new x86 backend, I think this should be OK; but if we feel differently, I can dig into this more and implement what is needed.

Because of that, the flag must be `false` by default to avoid breaking the legacy x86 backend. IMHO, we should flip it once we transition. Until then, we can ensure that we set it for specific embeddings that we control, e.g. in SpiderMonkey. Thoughts?

bnjbvr commented:
`CMOV_OVERFLOW` represents the initial bits for `cmov` with no cc set (well, it's because the condition code for overflow is 0), and ye olde recipe adds the cc bits correctly, when it encodes `selectif`. So I think it could be done without too much trouble in the old backend?

cfallin commented:
Just updated -- PTAL. Now the flag is on by default, and I managed to work out support for the legacy x86 backend as well -- by making our side-effecting-black-box-op `selectif_spectre_guard` rather than `select_spectre_guard`, we can just associate it with the same encoding as for `selectif` (except that no legalizations or optimizations understand the op so they will not alter it). I think this addresses all concerns above: on by default, works on all platforms, optimizations won't interfere. Thanks!

bnjbvr commented on cranelift/codegen/src/legalizer/heap.rs:
```diff
@@ -66,19 +66,14 @@ fn dynamic_addr(
     let bound = pos.ins().global_value(offset_ty, bound_gv);
-    let oob;
-    if access_size == 1 {
+    let cmp = if access_size == 1 {
```
nit: for symmetry with the other function, can you use a destructuring assignment here too? `let (cc, offset, bound) = ...`

cfallin replied:
Done!
